Adyen/Helpers/UIViewAnimation.swift

```diff
@@ -58,56 +62,67 @@ internal final class KeyFrameAnimationContext: AnimationContext {
+        base.animations.append(context)
+        
+        guard base.animations.count == 1 else { return }
+        animateNext(context: context)
     }
     
     /// :nodoc:
-    @objc internal func animateKeyframes(context: KeyFrameAnimationContext) {
-        if animationsMap.contains(context.animationKey) {
-            perform(#selector(animateKeyframes(context:)), with: context, afterDelay: 0.1)
-            return
-        }
-        animateKeyframesSynchronized(context: context)
+    public func animateKeyframes(context: KeyFrameAnimationContext) {
+        animate(context: context)
     }
     
-    @objc private func animateSynchronized(context: AnimationContext) {
-        animationsMap.insert(context.animationKey)
+    private func animateNext(context: AnimationContext?) {
+        if let context = context as? KeyFrameAnimationContext {
+            animateNextKeyframes(context: context)
+            return
+        }
+        guard let context = context else { return }
+        
+        func next() {
+            base.animations.removeFirst()
+            animateNext(context: base.animations.first)
+        }
+        
         UIView.animate(withDuration: context.duration,
                        delay: context.delay,
                        options: context.options,
                        animations: context.animations,
                        completion: {
                            context.completion?($0)
-                           self.animationsMap.remove(context.animationKey)
+                           next()
                        })
     }
     
-    @objc private func animateKeyframesSynchronized(context: KeyFrameAnimationContext) {
-        animationsMap.insert(context.animationKey)
+    private func animateNextKeyframes(context: KeyFrameAnimationContext?) {
+        guard let context = context else { return }
+        
+        func next() {
+            base.animations.removeFirst()
+            animateNext(context: base.animations.first)
+        }
         
         UIView.animateKeyframes(withDuration: context.duration,
                                 delay: context.delay,
                                 options: context.keyFrameOptions,
                                 animations: context.animations,
                                 completion: {
                                     context.completion?($0)
-                                    self.animationsMap.remove(context.animationKey)
+                                    next()
                                 })
     }
-    
+}
+
+private extension UIView {
     /// :nodoc:
-    private var animationsMap: Set<String> {
+    var animations: [AnimationContext] {
         get {
-            guard let value = objc_getAssociatedObject(self, &AssociatedKeys.animations) as? Set<String> else {
-                return Set<String>()
+            guard let value = objc_getAssociatedObject(self, &AssociatedKeys.animations) as? [AnimationContext] else {
```

Comment:
might be simplified as. i believe it's a common pattern, but not 100% sure
```swift
objc_getAssociatedObject(self, &AssociatedKeys.animations) as? [AnimationContext] ?? []
```

Reply:
👍